File: server/rest-api/model/adaclass.php
<?php
    class AdaClass {
        private $id;
        private $name;
        private $isFolderClass = false;
        private $isDocumentClass = false;
        private $description;
        private $properties;
        private $rights;
        private $parentClass;

        public function __construct($id, $name) {
            $this->id = $id;
            $this->name = $name;
            $this->properties = [];
            $this->rights = [];
        }

        public function getId() {
            return $this->id;
        }

        public function getName() {
            return $this->name;
        }

        public function setDescription($description) {
            $this->description = $description;
        }

        public function getDescription() {
            return $this->description;
        }

        public function setIsFolderClass($value) {
            $this->isFolderClass = $value;
        }

        public function isFolderClass() {
            return $this->isFolderClass==1;
        }

        public function setIsDocumentClass($value) {
            $this->isDocumentClass = $value;
        }

        public function isDocumentClass() {
            return $this->isDocumentClass==1;
        }

        public function addProperty($property) {
            $this->properties[sizeof($this->properties)] = $property;
        }

        public function getProperties() {
            return isset($this->properties) ? $this->properties: false;
        }

        public function setParentClass($id) {
            $this->parentClass = $id;
        }

        public function getParentClass() {
            return $this->parentClass;
        }

        public static function fromJson($json) {
            $result = new AdaClass(null, $json->name);
            if (isset($json->documentclass))
                $result->setIsDocumentClass($json->documentclass);
            if (isset($json->folderclass))
                $result->setIsFolderClass($json->folderclass);
            if (isset($json->parentclass))
                $result->setParentClass($json->parentclass);
            if (isset($result->properties)) {
                foreach($json->properties as $property)
                    $result->addProperty(Property::fromJson($property));
            }
            if (isset($json->rights)) {
                foreach($json->rights as $right) 
                    $result->addRight(GrantedRight::fromJson($right));
            }
            return $result;
        }

        public static function fromAddon($json) {
            $result = new AdaClass(null, $json->name);
            if (isset($json->documentclass))
                $result->setIsDocumentClass($json->documentclass);
            if (isset($json->folderclass))
                $result->setIsFolderClass($json->folderclass);
            if (isset($json->parentclass))
                $result->setParentClass($json->parentclassid);
            if (isset($json->description))
                $result->setDescription($json->description);
            if (isset($result->properties)) {
                foreach($json->properties as $property) {
                    $result->addProperty(Property::fromJson($property));
                }
            }
            return $result;
        }

        public function getRights() {
            return $this->rights;
        }

        public function addRight($right) {
            $this->rights[sizeof($this->rights)] = $right;
        }

        public static function toJson($classes) {
            if (is_array($classes)) {
                $json = '[';
                foreach($classes as $class) {
                    if ($json != '[')
                        $json .= ',';
                    $json .= AdaClass::toJson($class);
                }
                $json .= ']';
                return $json;
            }
            else {
                $json = '{';
                $json .= '"id":"' . $classes->getId() . '",';
                $json .= '"name":"' . $classes->getName() . '",';
                $json .= '"folderclass":' . ($classes->isFolderClass() ? 'true':'false') . ',';
                $json .= '"documentclass":' . ($classes->isDocumentClass() ? 'true': 'false');
                if (!is_null($classes->getParentClass())) {
                    $json .= ',"parentclass":"' . $classes->getParentClass() . '"';
                }
                if ($props = $classes->getProperties()) {
                    $json .= ',"properties":[';
                    $first = true;
                    foreach($props as $prop) {
                        $json .= ($first? '' : ',');
                        $first = false;
                        $json .= $prop->toJson();
                    }
                    $json .= ']';
                }
                if ($rights = $classes->getRights()) {
                    $json .= ',"grantedrights":[';
                    $first = true;
                    foreach($rights as $right) {
                        $json .= ($first? '': ',');
                        $first = false;
                        $json .= $right->toJson();
                    }
                    $json.=']';
                }
                $json .= '}';
                return $json;
            }
        }

        public function createObjectSchema() {
            $schema = "{";
            if ($this->getProperties() || $this->isDocumentClass()) {
                $schema .= "\"properties\": {\"type\": \"arrayofobjects\", \"schema\":{";
                $prefix = "";
                foreach($this->getProperties() as $property) {
                    $schema .= $prefix . "\"" . $property->getName() . "\": {";
                    $schema .= "\"type\":\"" . $property->getType() . "\",";
                    $schema .= "\"required\":" . ($property->isRequired() ? "true": "false");
                    $schema .= "}";
                    $prefix = ",";
                }
                $schema .= "}},";
            }
            if ($this->isDocumentClass()) {
                $schema .= "\"content\": {";
                $schema .= "\"schema\":\"content\",";
                $schema .= "\"required\": false";
                $schema .= "},";
            }
            $schema .= "\"rights\":{";
            $schema .= "\"type\":\"arrayofobjects\",";
            $schema .= "\"required\":false,";
            $schema .= "\"schema\": \"right\"";
            $schema .= "}";
            $schema .= "}";
            return $schema;
        }


        public function updateObjectSchema() {
            $schema = "{";
            $schema .= '"id": {"type": "string", "required":true},';
            $schema .= '"classid": {"type": "string", "required":false},';
            if ($this->getProperties() || $this->isDocumentClass()) {
                $schema .= "\"properties\": {\"arrayofobjects\", \"schema\":{";
                $prefix = "";
                foreach($this->getProperties() as $property) {
                    $schema .= $prefix . "\"" . $property->getName() . "\": {";
                    $schema .= "\"type\":\"" . $property->getType() . "\",";
                    $schema .= "\"required\":" . ($property->isRequired() ? "true": "false");
                    $schema .= "}";
                    $prefix = ",";
                }
                $schema .= "}},";
            }
            if ($this->isDocumentClass()) {
                $schema .= "\"content\": {";
                $schema .= "\"schema\":\"content\",";
                $schema .= "\"required\": false";
                $schema .= "},";
            }
            $schema .= "\"rights\":{";
            $schema .= "\"type\":\"arrayofobjects\",";
            $schema .= "\"required\":false,";
            $schema .= "\"schema\": \"right\"";
            $schema .= "}";
            $schema .= "}";
            return $schema;
        }

    }
?>

File: client/apis/ada-java-api/src/main/java/nl/arba/ada/client/api/AdaClass.java
package nl.arba.ada.client.api;

import com.fasterxml.jackson.annotation.JsonProperty;
import nl.arba.ada.client.api.exceptions.*;
import nl.arba.ada.client.api.security.GrantedRight;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Class representing an objectclass in a store
 */
public class AdaClass {
    private String id;
    private String name;
    private boolean folderClass;
    private boolean documentClass;
    private ArrayList <GrantedRight> rights = new ArrayList<>();
    private ArrayList <Property> properties = new ArrayList<>();
    private String description;

    private AdaClass parentClass;
    private String parentClassId;
    private Store store;

    /**
     * Set the unique id of the class
     * @param id The unique id of the class
     */
    public void setId(String id) {
        this.id = id;
    }

    /**
     * Get the unique id of the class
     * @return The unique id of the class
     */
    public String getId() {
        return id;
    }

    /**
     * Set the name of the class
     * @param value The name of the class
     */
    public void setName(String value) {
        name = value;
    }

    /**
     * Get the name of the class
     * @return The name of the class
     */
    public String getName() {
        return name;
    }

    /**
     * Set if the class is a folderclass
     * @param value <code>true</code> - class is a folderclass, <code>false</code> - the class is not a folderclass
     */
    @JsonProperty("folderclass")
    public void setFolderClass(boolean value) {
        folderClass = value;
    }

    /**
     * Get if the class is a folderclass
     * @return <code>true</code> - class is a folderclass, <code>false</code> - the class is not a folderclass
     */
    public boolean isFolderClass() {
        return folderClass;
    }

    /**
     * Set if the class is a documentclass
     * @param value <code>true</code> - class is a documentclass, <code>false</code> - the class is not a documentclass
     */
    @JsonProperty("documentclass")
    public void setDocumentClass(boolean value) {
        documentClass = value;
    }

    /**
     * Get if the class is a documentclass
     * @return <code>true</code> - class is a documentclass, <code>false</code> - the class is not a documentclass
     */
    public boolean isDocumentClass() {
        return documentClass;
    }

    /**
     * Add a granted right to the class
     * @param right The granted right
     * @see GrantedRight
     */
    public void addRight(GrantedRight right) {
        rights.add(right);
    }

    /**
     * Set the properties of the class
     * @param properties The properties
     * @see Property
     */
    public void setProperties(Property[] properties) {
        this.properties.clear();
        this.properties.addAll(Arrays.asList(properties));
    }

    /**
     * Add a property to the class
     * @param property The property to add
     * @see Property
     */
    public void addProperty(Property property) throws PropertyNotAddedException, InsufficientRightsException {
        if (this.getId() == null)
            properties.add(property);
        else {
            //Adding property to existing class
            getStore().getDomain().addProperty(getStore(), this, property);
        }
    }

    /**
     * Get the properties of the class
     * @return The properties of the class
     */
    public Property[] getProperties() {
        return properties.toArray(new Property[0]);
    }

    public boolean hasProperty(String name) {
        return properties.stream().anyMatch(p -> p.getName().equals(name));
    }

    public Property getProperty(String name) throws PropertyNotFoundException {
        if (hasProperty(name))
            return properties.stream().filter(p ->p.getName().equals(name)).findFirst().get();
        else
            throw new PropertyNotFoundException();
    }

    public Property editProperty(Property changed) throws PropertyNotFoundException, PropertyNotChangedException, InsufficientRightsException {
        if (properties.stream().anyMatch(p -> p.getId().equals(changed.getId()))) {
            getStore().getDomain().editProperty(getStore(), this, changed);
            try {
                AdaClass refreshedClass = getStore().getAdaClass(getId());
                properties.clear();
                properties.addAll(Arrays.asList(refreshedClass.getProperties()));
            }
            catch (Exception err) {}
            return null;
        }
        else
            throw new PropertyNotFoundException();
    }

    public void deleteProperty(Property todelete) throws InsufficientRightsException, PropertyNotDeletedException{
        getStore().getDomain().deleteProperty(getStore(), this, todelete);
    }

    /**
     * Convert to property to a json representation
     * @return A json representation of the property
     */
    public String toJson() {
        return "{\"name\" : \"" + getName() + "\", " +
            "\"documentclass\": " + documentClass + "," +
            "\"folderclass\": " + folderClass + "," +
            (getParentClass() == null ? "": ("\"parentclass\": \"" + getParentClass().getId() + "\",")) +
            "\"rights\":[" +
                rights.stream().map(r -> r.toJson()).collect(Collectors.joining(",")) +
            "]," +
            "\"properties\":[" +
                properties.stream().map(p ->p.toJson()).collect(Collectors.joining(",")) +
            "]" +
            "}";
    }

    /**
     * Set the description of the class
     * @param description The description
     */
    public void setDescription(String description) {
        this.description = description;
    }

    /**
     * Get the description of the class
     * @return The description
     */
    public String getDescription() {
        return description;
    }

    /**
     * Set the rights that applies to the class
     * @param rights The rights that applies to the class
     */
    public void setGrantedrights(GrantedRight[] rights) {
        this.rights.clear();
        this.rights.addAll(Arrays.asList(rights));
    }

    public List<GrantedRight> getGrantedRights() {
        return rights;
    }

    public void setParentClass(AdaClass parentclass) {
        this.parentClass = parentclass;
    }

    public void setParentclass(String classid) {
        parentClassId = classid;
    }

    public String getParentClassId() {
        return parentClassId;
    }

    public AdaClass getParentClass() {
        if (parentClass == null && parentClassId != null && getStore() != null) {
            try {
                parentClass = getStore().getAdaClass(parentClassId);
            }
            catch (AdaClassNotFoundException cnfe) {}
        }
        return parentClass;
    }

    public Store getStore() {
        return store;
    }

    public void setStore(Store store) {
        this.store = store;
    }

    public void update() throws InsufficientRightsException, LostRightsException{
        getStore().getDomain().updateClass(this);
    }
}
